mart/attack/adversary.py

```diff
@@ -6,338 +6,116 @@
-                )
-                self.examine(input=input, target=target, model=model, **kwargs)
-                self.on_examine_end(
-                    adversary=self, input=input, target=target, model=model, **kwargs
-                )
-
-                # Check the done condition here, so that every update of perturbation is examined.
-                if not self.done:
-                    self.on_advance_start(
-                        adversary=self,
-                        input=input,
-                        target=target,
-                        model=model,
-                        **kwargs,
-                    )
-                    self.advance(
-                        input=input,
-                        target=target,
-                        model=model,
-                        **kwargs,
-                    )
-                    self.on_advance_end(
-                        adversary=self,
-                        input=input,
-                        target=target,
-                        model=model,
-                        **kwargs,
-                    )
-                    # Update cur_iter at the end so that all hooks get the correct cur_iter.
-                    self.cur_iter += 1
-                else:
-                    break
-            except StopIteration:
-                break
-
-        self.on_run_end(adversary=self, input=input, target=target, model=model, **kwargs)
-
-    # Make sure we can do autograd.
-    # Earlier Pytorch Lightning uses no_grad(), but later PL uses inference_mode():
-    #   https://github.com/Lightning-AI/lightning/pull/12715
-    @torch.enable_grad()
-    @torch.inference_mode(False)
-    def examine(
-        self,
-        *,
-        input: torch.Tensor | tuple,
-        target: torch.Tensor | dict[str, Any] | tuple,
-        model: torch.nn.Module,
-        **kwargs,
-    ):
-        """Examine current perturbation, update self.gain and self.found."""
-
-        # Clone tensors for autograd, in case it was created in the inference mode.
-        # FIXME: object detection uses non-pure-tensor data, but it may have cloned somewhere else implicitly?
-        if isinstance(input, torch.Tensor):
-            input = input.clone()
-        if isinstance(target, torch.Tensor):
-            target = target.clone()
+        self.attacker = attacker
+
+        if self.attacker is None:
+            # Enable attack to be late bound in forward
+            self.attacker = partial(
+                pl.Trainer,
+                num_sanity_val_steps=0,
+                logger=False,
+                max_epochs=0,
+                limit_train_batches=kwargs.pop("max_iters", 10),
+                callbacks=list(kwargs.pop("callbacks", {}).values()),  # dict to list of values
+                enable_model_summary=False,
+                enable_checkpointing=False,
+                enable_progress_bar=False,
```

Comment:
I think we will fix the progress bar in a separate PR.

Reply:
Yeah it's hard to fix because you have to get the `position` correct since there are multiple nested progress bars.